packages/odyssey-react-mui/src/ui-shell/createMessageBus.ts

```diff
@@ -0,0 +1,50 @@
+/*!
+ * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+export type MessageEvent<Message> = (message: Message) => void;
+
+export type PublishMessage<Message> = MessageEvent<Message>;
+
+export type UnsubscribeMessageSubscription = () => void;
+
+export type MessageSubscription<Message> = (
+  subscription: MessageEvent<Message>,
+) => UnsubscribeMessageSubscription;
+
+export type MessageBus<Message> = {
+  publish: PublishMessage<Message>;
+  subscribe: MessageSubscription<Message>;
+};
+
+export const createMessageBus = <Message = void>(): MessageBus<Message> => {
+  const subscribers = new Map<symbol, MessageEvent<Message>>();
+
+  const publish: PublishMessage<Message> = (message) => {
+    Array.from(subscribers.values()).forEach((subscriber) => {
+      subscriber(message);
+    });
+  };
+
+  const subscribe: MessageSubscription<Message> = (subscription) => {
+    const subscriptionId = Symbol();
+    subscribers.set(subscriptionId, subscription);
+
+    return () => {
+      subscribers.delete(subscriptionId);
+    };
+  };
+
+  return {
+    publish,
+    subscribe,
+  };
+};
```

Comment:
Have you considered BroadcastChannel for this? https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API

Reply:
## My original idea
I considered using a DOM element as the event bus as it has a built-in way to setup and remove listeners as well as `postMessage`, but I decided against it because I wanted something quick and simple.

Using something native to the browser would save on code and custom logic, but I'd have to wrap `addEventListener` anyway, and I'd have to give them the wrapped postMessage in `publish()`. So from my perspective, I thought that adding that unnecessary complexity would be something people call out in the PR, and I'd remove it anyway.

That was my reasoning at least.

## Broadcast Channel
Broadcast Channel looks pretty cool! I think it'd run into the same issue as an DOM element though because I'd have to ultimately wrap it to make it work with the implementation we have. I'm also not sure how Broadcast Channels handle sending React components and objects in memory.

## What about iframes?
One thing I wanted to avoid is anything in the global scope, but Access Certification renders in an iframe. To communicate with the nav, it would need a Broadcast Channel. To use one, we'd need to export that type to their app, and the types can get outta sync with the JSP.

There are a lot of issues, but it's possible to build Broadcast Channel into the current implementation without changing the API.

Reply:
For now, I'd recommend not worrying how consumers handle cross-context communications. We'd hope to encourage them to use Tightbeam, but really how they do it shouldn't be built into this impl IMO